### PR TITLE
Helmclient metrics

### PR DESCRIPTION
**Pull request by tomahawk28**
- Adding error gauge, the time duration for each operation on helmclient.

### CODE REVIEW

File: helmclient.go [hunk 1]
```diff
@@ -118,7 +119,9 @@ func New(config Config) (*Client, error) {
 }
 
 // DeleteRelease uninstalls a chart given its release name.
-func (c *Client) DeleteRelease(ctx context.Context, releaseName string, options ...helmclient.DeleteOption) error {
+func (c *Client) DeleteRelease(ctx context.Context, releaseName string, options ...helmclient.DeleteOption) (err error) {
```
Comment (tomahawk28) [inline helmclient.go]: I named on returning `error` variable so we can check it whether it's nil or not when we exit this function.

File: helmclient.go [hunk 2]
```diff
@@ -118,7 +119,9 @@ func New(config Config) (*Client, error) {
...
+	eventName := "delete_release"
```
Comment (tomahawk28) [inline helmclient.go]: I'm a quite python fan, but not sure what is `__func__` equivalent for golang.

File: helmclient.go [hunk 3]
```diff
@@ -574,85 +597,107 @@ func (c *Client) ListReleaseContents(ctx context.Context) ([]*ReleaseContent, er
...
 
-	resChan, errChan := c.newHelmClientFromTunnel(t).RunReleaseTest(releaseName, helmclient.ReleaseTestTimeout(int64(runReleaseTestTimout)))
+	resChan, errChan := c.newHelmClientFromTunnel(tunnel).RunReleaseTest(releaseName, helmclient.ReleaseTestTimeout(int64(runReleaseTestTimout)))
 	if IsReleaseNotFound(err) {
 		return backoff.Permanent(microerror.Mask(err))
 	} else if err != nil {
 		return microerror.Mask(err)
 	}
 
+W:
```
Comment (tomahawk28) [inline helmclient.go]: Adding the label for exiting from the infinite loop.

File: helmclient.go (hunk 2 above)
Comment (tomahawk28) [reply helmclient.go]: Should it be const strings?

File: metric.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
+package helmclient
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	PrometheusNamespace = "operatorkit"
```
Comment (rossf7) [inline metric.go]: Should this be `helmclient`?

File: helmclient.go [hunk 5]
```diff
@@ -574,85 +597,107 @@ func (c *Client) ListReleaseContents(ctx context.Context) ([]*ReleaseContent, er
...
 	for {
 		select {
 		case err := <-errChan:
 			if err != nil {
-				return microerror.Mask(err)
+				err = microerror.Mask(err)
+				break W
```
Comment (rossf7) [inline helmclient.go]: Why is this `break W`?

File: helmclient.go (hunk 2 above)
Comment (rossf7) [reply helmclient.go]: I'm not sure either. Maybe the `reflect` package? I'd suggest researching it if you want to use this.

File: helmclient.go (hunk 1 above)
Comment (kopiczko) [reply helmclient.go]: I still don't get why. AFAICT this should be used only for documentation purposes.

I also don't think `errorReporting` is reliable. I already see some errors redeclared in backoff functions for example.

I'd rather go with `deleteRelease` called by `DeleteRelease` and check for errors there.

File: metric.go (hunk 4 above)
Comment (tomahawk28) [reply metric.go]: 🤦‍♂ Oh, yes.

File: helmclient.go (hunk 5 above)
Comment (tomahawk28) [reply helmclient.go]: Otherwise, it would only break inner `select` statement.

File: helmclient.go [hunk 6]
```diff
@@ -361,6 +361,14 @@ func (c *Client) EnsureTillerInstalledWithValues(ctx context.Context, values []s
 // values provided when the chart was installed. The releaseName is the name
 // of the Helm Release that is set when the Helm Chart is installed.
 func (c *Client) GetReleaseContent(ctx context.Context, releaseName string) (*ReleaseContent, error) {
+	releaseContent, err := c.getReleaseContent(ctx, releaseName)
```
Comment (tomahawk28) [inline helmclient.go]: I follow Pawel's suggestion to check the error result and update metrics.

File: Gopkg.toml [hunk 7]
```diff
@@ -107,3 +107,7 @@ required = [
 [prune]
   go-tests = true
   unused-packages = true
+
+[[constraint]]
+  name = "github.com/prometheus/client_golang"
+  version = "0.9.3"
```
Comment (rossf7) [inline Gopkg.toml]: This is in the wrong place.

File: helmclient.go [hunk 8]
```diff
@@ -361,6 +362,17 @@ func (c *Client) EnsureTillerInstalledWithValues(ctx context.Context, values []s
 // values provided when the chart was installed. The releaseName is the name
 // of the Helm Release that is set when the Helm Chart is installed.
 func (c *Client) GetReleaseContent(ctx context.Context, releaseName string) (*ReleaseContent, error) {
+	eventName := "get_release_content"
```
Comment (rossf7) [inline helmclient.go]: I find this hard to read. Please split up the lines with whitespace.

File: helmclient.go [hunk 9]
```diff
@@ -3,6 +3,7 @@ package helmclient
 import (
 	"context"
 	"fmt"
+	"github.com/prometheus/client_golang/prometheus"
```
Comment (rossf7) [inline helmclient.go]: Needs ordering.

File: metric.go [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+			Help:      "Number of helmclient errors.",
+		},
+		[]string{"event"},
+	)
+	controllerHistogram = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: PrometheusNamespace,
+			Subsystem: PrometheusSubsystem,
+			Name:      "event",
+			Help:      "Histogram for events within the helmclient controller.",
```
Comment (rossf7) [inline metric.go]:
```suggestion
			Help:      "Histogram for events within the helmclient library.",
```

File: helmclient.go [hunk 11]
```diff
@@ -361,6 +362,21 @@ func (c *Client) EnsureTillerInstalledWithValues(ctx context.Context, values []s
 // values provided when the chart was installed. The releaseName is the name
 // of the Helm Release that is set when the Helm Chart is installed.
 func (c *Client) GetReleaseContent(ctx context.Context, releaseName string) (*ReleaseContent, error) {
+	eventName := "get_release_content"
+
+	t := prometheus.NewTimer(controllerHistogram.WithLabelValues(eventName))
+
+	releaseContent, err := c.getReleaseContent(ctx, releaseName)
+	if err != nil {
+		controllerErrorGauge.WithLabelValues(eventName).Inc()
```
Comment (rossf7) [inline helmclient.go]: I think the error masking won't work here and then the stack traces will be wrong. 

After updating the gauge I think we should return the masked error like we usually do.

File: helmclient.go [hunk 12]
```diff
@@ -361,6 +362,21 @@ func (c *Client) EnsureTillerInstalledWithValues(ctx context.Context, values []s
...
+	}
+
+	t.ObserveDuration()
+
+	return releaseContent, err
```
Comment (rossf7) [inline helmclient.go]:
```suggestion
	return releaseContent, nil
```

File: helmclient.go (hunk 11 above)
Comment (tomahawk28) [reply helmclient.go]: Good catch, never thought about that.

File: helmclient.go [hunk 13]
```diff
@@ -361,6 +362,22 @@ func (c *Client) EnsureTillerInstalledWithValues(ctx context.Context, values []s
...
 // of the Helm Release that is set when the Helm Chart is installed.
 func (c *Client) GetReleaseContent(ctx context.Context, releaseName string) (*ReleaseContent, error) {
+	eventName := "get_release_content"
+
+	t := prometheus.NewTimer(histogram.WithLabelValues(eventName))
+
+	releaseContent, err := c.getReleaseContent(ctx, releaseName)
+	if err != nil {
+		errorGauge.WithLabelValues(eventName).Inc()
+		return releaseContent, microerror.Mask(err)
```
Comment (rossf7) [inline helmclient.go]:
```suggestion
		return nil, microerror.Mask(err)
```

File: helmclient.go [hunk 14]
```diff
@@ -361,6 +362,22 @@ func (c *Client) EnsureTillerInstalledWithValues(ctx context.Context, values []s
...
+	}
+
+	t.ObserveDuration()
```
Comment (JosephSalisbury) [inline helmclient.go]: Why not defer? Then you can keep it closer to the timer creation

**Pull request merged**